### PR TITLE
Make blocksize a local variable

### PR DESCRIPTION
**Pull request by ncanceill**
# Make `blocksize` a local variable
## The bug

In the main file `src/frag_find.cpp`, the block size is currently handled as a global variable:

``` c++
uint32_t blocksize = DEFAULT_BLOCKSIZE;
```

This introduces a bug always reproducible with the standard build `./bootstrap.sh && ./configure && make all`.
### Status

Working on:
- Darwin 12.3 x86_64 (Mac OS 10.8) with g++ 4.2.1
- Linux 3.2 x86_64 (Ubuntu 12.04) with g++ 4.6.3
### Description

Using any block size but 512 bytes leads to inconsistent results
### How to reproduce

Using block size 512:
- Scan something containing a blank block (512 consecutive zero bytes) for a master file (of any size at least the block size) full of zeros.
- Expected/actual behavior: the block size is 512; the blank block is detected.

Using any block size between 4 and 511:
- Scan the same thing containing a blank block (512 consecutive zero bytes) for a master file (of any size at least the block size) full of zeros.
- Expected behavior: at least a blank block is detected.
- Actual behavior: the target (the scanned image file) gets the proper block size, but the master file and the reports get blocksize 512 no matter what; nothing is detected even when the master file is larger than 512 bytes.
## The fix

I fixed it in the brutal simple way: make `blocksize` a local variable everywhere.
### Status

`uint32_t blocksize` is no longer declared in the global scope. Instead, it is now declared in `main()`, and set to `DEFAULT_BLOCKSIZE` like before. After that, it may be set in the option-parsing process like before.

Finally, it has been included as a parameter in all the function signatures where it was needed, all directly or undirectly called from `main()`.
### Side-effects

`blocksize` is actually needed in `read_md5deep`, which used to get called during the option-parsing process. To avoid inconsistencies that could occur if option `-m` was used before option `-b`, an intermediate local variable is now used for `-m` during the option-parsing process.

### CODE REVIEW

File: src/frag_find.cpp [hunk 1]
```diff
@@ -550,6 +551,7 @@ int main(int argc,char **argv)
 	fprintf(stderr,"M must be between 5 and 32.\n");
 	exit(1);
     }
+    if (!opt_md5.empty()) masters.read_md5deep(opt_md5.c_str(),blocksize);
```
Comment (ncanceill) [inline src/frag_find.cpp:554]: Due to a side-effect of fixing #1, calling `masters.read_md5deep()` during option parsing could lead to inconsistencies depending on the order of option parsing, so let's call it afterwards using a local variable `opt_md5`.

File: src/frag_find.cpp [hunk 2]
```diff
@@ -59,7 +59,6 @@
 void *initial_break = 0;
 
 blockfile imagefile;			// the file we are reading
-uint32_t blocksize = DEFAULT_BLOCKSIZE;
```
Comment (ncanceill) [inline src/frag_find.cpp:62]: Having `blocksize` as a global variable leads to inconsistencies mentioned in #1.

**Comment by simsong** [comment]
I really don't want to take pull requests that introduce bugs. I'll see about validating the claimed bug with different blocksizes and look into a cleaner fix. Can you provide a simple test case demonstrating the bug?

**Comment by ncanceill** [comment]
Totally granted, as being a file-wide modification that affects the API itself, this pull request does not exactly abide by common guidelines.

However, neither does the overuse of global variables in C++ because they have **program-wide scope** and can be externalized. This is a powerful feature, but it can make debugging and refactoring very hard; also, it can easily introduce bugs as this one.

As mentioned, when the `main()` function calls master file functions like `load()`, the default `blocksize`is used despite the `-b` option. This indicates either a scope limitation in the calls from `main()`, or a problem of value updating...

To sum up:
- The bug is complex and takes various forms.
- Switching to a local variable is a clear fix.

So the trade-off you want to consider is: investigating a deep bug, or taking the risk of my large refactoring.

Let me provide you with a simple sample [script](https://gist.github.com/ncanceill/5756997) to reproduce. Maybe the best compromise is for me to conduct some more tests on my pull's robustness.
